---
title: "Visualizing STR Loci and Overlaps"
author: "Laurel Hiatt"
date: "2023-05-24"
output: html_document
---

```{r setup, include=FALSE}
knitr::opts_chunk$set(echo = TRUE)
gnomADSTRcalls = read.csv('/Users/quinlan/Documents/Quinlan-PhD/UDN+STRdb/gnomAD_STR_genotypes__2022_01_20.tsv', sep = '\t', stringsAsFactors = FALSE)

```

## R Markdown

This is an R Markdown document. Markdown is a simple formatting syntax for authoring HTML, PDF, and MS Word documents. For more details on using R Markdown see <http://rmarkdown.rstudio.com>.

When you click the **Knit** button a document will be generated that includes both content as well as the output of any embedded R code chunks within the document. You can embed an R code chunk like this:

```{r libraries}
library(Gviz)
#library(GenomicRanges) req'd pkg for Gviz
library(rtracklayer)
# library(IRanges) req'd pkg for Gviz
library(tidyverse)

# extrasmallgnomADSTR <- read.csv("/Users/quinlan/Documents/Quinlan-PhD/UDN+STRdb/gnomAD_STR_genotypes__2022_01_20_noheader_sorted_extrasmall.bed", sep = "\t", header = FALSE, stringsAsFactors = F)


intersection_wiggle <- read.csv("/Users/quinlan/Documents/Quinlan-PhD/UDN+STRdb/intersectionofgnomADandrepeats_wiggle.txt", sep = "\t", header = FALSE, stringsAsFactors = F)

# Get unique values from column 4
unique_values <- unique(intersection_wiggle$V4)

# Create new dataframe objects dynamically
for (value in unique_values) {
  subset_df <- intersection_wiggle[intersection_wiggle$V4 == value, ]
  assign(paste0("df_", value), subset_df)
}


# Define a function to create a BED file from a dataframe
create_RE_bed <- function(df) {
  bed_df <- df[, c("V7", "V8", "V9", "V21")]
  colnames(bed_df) <- c("chrom", "start", "end", "id")
  return(bed_df)
}

# create_reads_bed <- function(df) {
#   bed_df <- df[, c("V1", "V2", "V3")]
#   colnames(bed_df) <- c("chrom", "start", "end")
#   return(bed_df)
# }

create_lociID_bed <- function(df) {
  bed_df <- separate(df, V6, into = c("chrom", "start", "end"), sep = "[:-]")
  bed_df <- unique(bed_df[, c("chrom", "start", "end")])
  return(bed_df)
}


# Get unique values from "Id" column
ids <- unique(df_TCF4$V22)

# Create a color palette based on unique Ids
color_palette <- rainbow(length(ids), start = 0.7, end = 0.8)

# Create a named vector to map Ids to colors
id_colors <- setNames(color_palette, ids)

idtrack <- AnnotationTrack(makeGRangesFromDataFrame(create_lociID_bed(df_TCF4)), name = "RefRegion", fill = "black")
#overlaptrack <- AnnotationTrack(makeGRangesFromDataFrame(create_reads_bed(df_TCF4)), name = "Overlap")
atrack <- AnnotationTrack(makeGRangesFromDataFrame(create_RE_bed(df_TCF4)), 
                          id = df_TCF4$V22, 
                          fill = id_colors[df_TCF4$V22], 
                          showFeatureId=TRUE)
gtrack <- GenomeAxisTrack()
itrack <- IdeogramTrack(genome = "hg38", chromosome = unique(df_TCF4$V1))
plotTracks(trackList = list(itrack, gtrack, idtrack, atrack), main = "TCF4")

```


``` {r loop with names}
intersection_wiggle <- read.csv("/Users/quinlan/Documents/Quinlan-PhD/UDN+STRdb/intersectionofgnomADandrepeats_wiggle.txt", sep = "\t", header = FALSE, stringsAsFactors = FALSE)

# Get unique values from column 4
unique_values <- unique(intersection_wiggle$V4)

# Create a list to store the dataframes
dataframes <- list()

# Create new dataframe objects dynamically
for (value in unique_values) {
  subset_df <- intersection_wiggle[intersection_wiggle$V4 == value, ]
  df_name <- paste0("df_", value)
  dataframes[[df_name]] <- subset_df
}

# Define a function to create a BED file from a dataframe
create_RE_bed <- function(df) {
  bed_df <- df[, c("V7", "V8", "V9", "V21")]
  colnames(bed_df) <- c("chrom", "start", "end", "id")
  return(bed_df)
}

# Define a function to create loci ID bed
create_lociID_bed <- function(df) {
  bed_df <- separate(df, V6, into = c("chrom", "start", "end"), sep = "[:-]")
  bed_df <- unique(bed_df[, c("chrom", "start", "end")])
  return(bed_df)
}

processDataframes <- function(dataframes) {
  for (i in seq_along(dataframes)) {
    df <- dataframes[[i]]
    
    # Create a unique identifier for the dataframe
    df_id <- sub("^df_", "", names(dataframes)[i])
    
    # Create the BED file from the dataframe
    bed_df <- create_RE_bed(df)
    
    # Get unique values from the identifier column
    ids <- unique(df$V22)
    
    # Create a color palette based on unique ids
    color_palette <- colorRampPalette(c("blue", "green"))(length(ids))
    
    # Create a named vector to map ids to colors
    id_colors <- setNames(color_palette, ids)
    
    # Create the tracks for plotting
    idtrack <- AnnotationTrack(makeGRangesFromDataFrame(create_lociID_bed(df)), name = "RefRegion", fill = "black")
    atrack <- AnnotationTrack(makeGRangesFromDataFrame(bed_df), id = df$V22, fill = id_colors[df$V22], showFeatureId = TRUE)
    gtrack <- GenomeAxisTrack()
    itrack <- IdeogramTrack(genome = "hg38", chromosome = unique(df$V1))
    
    # Create the plot
    plotTracks(trackList = list(itrack, gtrack, idtrack, atrack), main = df_id)
  }
}

# Usage:
processDataframes(dataframes)
```


```{r yesnomaybe}
decision <- data.frame(
  stringsAsFactors = FALSE,
                          Loci = c("AFF2",
                                   "AR","ARX_1","ARX_2","ATN1","ATXN1",
                                   "ATXN10","ATXN2","ATXN3","ATXN7","ATXN8OS",
                                   "BEAN1","C9ORF72","CACNA1A","CNBP","COMP",
                                   "CSTB","DAB1","DIP2B","DMD","DMPK",
                                   "EIF4A3","FMR1","FOXL2","FXN","GIPC1","GLS",
                                   "HOXA13_1","HOXA13_2","HOXA13_3",
                                   "HOXD13","HTT","JPH3","LRP12","MARCHF6","NIPA1",
                                   "NOP56","NOTCH2NLC","NUTM2B-AS1",
                                   "PABPN1","PHOX2B","PPP2R2B","PRDM12","PRNP",
                                   "RAPGEF2","RFC1","RUNX2","SAMD12","SOX3",
                                   "STARD7","TBP","TBX1","TCF4","TNRC6A",
                                   "VWA1","XYLT1","YEATS2","ZIC2","ZIC3"),
                 Joint.Verdict = c("Maybe",
                                   "Yes","No!!!!","No!!!!","Yes","Yes",
                                   "Yes","Yes","No!!!!","Maybe","Yes","Maybe",
                                   "Yes","Yes","Yes","Yes","Yes","Maybe",
                                   "No!!!!","Maybe","Yes","Yes","Yes",
                                   "No!!!!","Yes","Yes","Yes","No!!!!","No!!!!",
                                   "No!!!!","No!!!!","Yes","Yes","Yes",
                                   "No!!!!","Maybe","Yes","No!!!!","Yes",
                                   "Maybe","No!!!!","Maybe","Yes","Yes",
                                   "No!!!!","Maybe","No!!!!","No!!!!","No!!!!",
                                   "Maybe","Yes","No!!!!","Yes","No!!!!","Yes",
                                   "Maybe","No!!!!","No!!!!","No!!!!")
            )

```

``` {r save as files}
processDataframes <- function(dataframes) {
  for (i in seq_along(dataframes)) {
    df <- dataframes[[i]]
    
    # Create a unique identifier for the dataframe
    df_id <- sub("^df_", "", names(dataframes)[i])
    
    # Create the BED file from the dataframe
    bed_df <- create_RE_bed(df)
    
    # Get unique values from the identifier column
    ids <- unique(df$V22)
    
    # Create a color palette based on unique ids
    color_palette <- colorRampPalette(c("blue", "green"))(length(ids))
    
    # Create a named vector to map ids to colors
    id_colors <- setNames(color_palette, ids)
    
    # Create the tracks for plotting
    idtrack <- AnnotationTrack(makeGRangesFromDataFrame(create_lociID_bed(df)), name = "RefRegion", fill = "black")
    atrack <- AnnotationTrack(makeGRangesFromDataFrame(bed_df), id = df$V22, fill = id_colors[df$V22], showFeatureId = TRUE)
    gtrack <- GenomeAxisTrack()
    itrack <- IdeogramTrack(genome = "hg38", chromosome = unique(df$V1))
    
    # Create the plot
    plotTracks(trackList = list(itrack, gtrack, idtrack, atrack), main = df_id)
    
    # Create a folder based on the verdict
    verdict <- decision$verdict[1]
    folder_path <- file.path(verdict)
    dir.create(folder_path, showWarnings = FALSE, recursive = TRUE)
    
    # Save the plot in the corresponding folder
    plot_file <- paste0(df_id, ".pdf")
    plot_path <- file.path(folder_path, plot_file)
    pdf(plot_path)
    plotTracks(trackList = list(itrack, gtrack, idtrack, atrack), main = df_id)
    dev.off()
  }
}

# Usage:
processDataframes(dataframes)

```

```{r verdict folder}
processDataframes <- function(dataframes, verdict_df) {
  for (i in seq_along(dataframes)) {
    df <- dataframes[[i]]
    df_name <- names(dataframes)[i]
    
    # Remove "df_" prefix from dataframe name
    df_id <- sub("^df_", "", df_name)
    
    # Check if the dataframe name exists in the verdict dataframe
    if (df_id %in% verdict_df$Loci) {
      # Get the corresponding verdict for the dataframe
      verdict <- verdict_df$Joint.Verdict[verdict_df$Loci == df_id]
      
      # Create the BED file from the dataframe
      bed_df <- create_RE_bed(df)
      
      # Get unique values from the identifier column
      ids <- unique(df$V22)
      
      # Create a color palette based on unique ids
      color_palette <- colorRampPalette(c("blue", "green"))(length(ids))
      
      # Create a named vector to map ids to colors
      id_colors <- setNames(color_palette, ids)
      
      # Create the tracks for plotting
      idtrack <- AnnotationTrack(makeGRangesFromDataFrame(create_lociID_bed(df)), name = "RefRegion", fill = "black")
      atrack <- AnnotationTrack(makeGRangesFromDataFrame(bed_df), id = df$V22, fill = id_colors[df$V22], showFeatureId = TRUE)
      gtrack <- GenomeAxisTrack()
      itrack <- IdeogramTrack(genome = "hg38", chromosome = unique(df$V1))
      
      # Create the plot
      plotTracks(trackList = list(itrack, gtrack, idtrack, atrack), main = df_id)
      
      # Create a folder based on the verdict
      folder_path <- file.path(verdict)
      dir.create(folder_path, showWarnings = FALSE, recursive = TRUE)
      
      # Save the plot in the corresponding folder
      plot_file <- paste0(df_id, ".pdf")
      plot_path <- file.path(folder_path, plot_file)
      pdf(plot_path)
      plotTracks(trackList = list(itrack, gtrack, idtrack, atrack), main = df_id)
      dev.off()
    }
  }
}

# Usage:
processDataframes(dataframes, decision)

```


## Not Working code



```{r bleh}


# # Iterate over the dataframes and create BED files
# for (value in unique_values) {
#   # Get the corresponding dataframe
#   df <- get(paste0("df_", value))
#   
#   # Create the BED file
#   bed_filename <- paste0("bed_", value, ".bed")
#   create_bed_file(df, bed_filename)
#   
#   # Create the Gviz plot using the BED file
#   track <- DataTrack(range = bed_filename, genome = "hg38")
#   gtrack <- GenomeAxisTrack()
#   plotTracks(trackList = list(gtrack, track))
# }
# 
# 
# bed_df <- df_AFF2[, c("V7", "V8", "V9", "V21")]
# colnames(bed_df) <- c("chrom", "start", "end", "id")

### Stuff that didn't work
# bed_extrasmallgnomADSTR <- extrasmallgnomADSTR[,1:3]
# 
# # Replace 'bed_file_path' with the path to your BED file
# bed_file_path <- "/Users/quinlan/Documents/Quinlan-PhD/UDN+STRdb/gnomAD_STR_genotypes__2022_01_20_noheader_sorted_extrasmall.bed"
# 
# # Specify the columns to import (1-based indices)
# columns <- c(1, 2, 3)
# 
# # Import the selected columns from the BED file into a GenomicRanges object
# 
# colnames(bed_extrasmallgnomADSTR) <- c("chrom", "start", "end")
# 
# bed00 <-  makeGRangesFromDataFrame(bed_extrasmallgnomADSTR)
# #Annotation track, title ="pARmm19"
# 
# 
# 
# atrack <- AnnotationTrack(bed00, name = "gnomAD_STR")
# #plotTracks(atrack)
# gtrack <- GenomeAxisTrack()
# #plotTracks(list(gtrack, atrack))
# itrack <- IdeogramTrack(genome = "hg38", chromosome = "chr7")
# 
# plotTracks(gtrack, itrack,  from = 26700000, to = 26750000)
# 
# 
# 
# 
# 
# for (i in 1:length(bed00)) {
#   # Extract the current locus
#   locus <- bed00[i, ]
# 
#   # Create a plot for the current locus
#   plot_track <- DataTrack(start = locus$V2, end = locus$V3)
#   # Create the plot with the genomic region track and the current locus track
#   plot <- plotTracks(trackList = list(gtrack, plot_track), from = locus$V2, to = locus$V3)
# 
#   # Save the plot to a file
#   filename <- paste0("plot_", i, ".pdf")
#   pdf(filename)
#   print(plot)
#   dev.off()
# }

```

Note that the `echo = FALSE` parameter was added to the code chunk to prevent printing of the R code that generated the plot.
